### PR TITLE
Implement Status for Application CRD

### PR DESCRIPTION
**Pull request by barney-s**
Implemented a draft version of Status for application crd.

Reference:
https://github.com/kubernetes-sigs/application/pull/77
https://github.com/kubernetes-sigs/application/pull/79 (closest to this)
https://github.com/kubernetes-sigs/application/issues/78

1. Manually vendor kubesdk
2. Use generic reconciler in Application controller loop.
3. Hack Mutate to return observed as expected (need a slightly better way)
4. Use meta.Status

Known Issues:
1. Need to ReSync with open proposals once they mature.
2. Periodic syncs(2m) to update status. Need to add dynamic watchers.
3. Unknown resources need to be handled.

Example Status with these changes:
```yaml
status:
    components:
    - link: /apis/apps/v1beta2/namespaces/default/statefulsets/mc-base-nfs
      name: mc-base-nfs
      status: Ready
    - link: /apis/apps/v1beta2/namespaces/default/statefulsets/mc-base-mysql
      name: mc-base-mysql
      status: Ready
    - link: /apis/apps/v1beta1/namespaces/default/statefulsets/mc-base-mysql
      name: mc-base-mysql
      status: Ready
    - link: /apis/apps/v1beta1/namespaces/default/statefulsets/mc-base-nfs
      name: mc-base-nfs
      status: Ready
    - link: /apis/apps/v1/namespaces/default/statefulsets/mc-base-mysql
      name: mc-base-mysql
      status: Ready
      sts:
        currentcount: 1
        progress: 0
        readycount: 1
        replicas: 1
    - link: /apis/apps/v1/namespaces/default/statefulsets/mc-base-nfs
      name: mc-base-nfs
      status: Ready
      sts:
        currentcount: 1
        progress: 0
        readycount: 1
        replicas: 1
    - link: /api/v1/namespaces/default/services/mc-base-sql
      name: mc-base-sql
      status: Ready
    - link: /api/v1/namespaces/default/services/mc-base-nfs
      name: mc-base-nfs
      status: Ready
    - link: /apis/policy/v1beta1/namespaces/default/poddisruptionbudgets/mc-base-mysql
      name: mc-base-mysql
      pdb:
        currenthealthy: 1
        desiredhealthy: 1
      status: Ready
    - link: /apis/policy/v1beta1/namespaces/default/poddisruptionbudgets/mc-base-nfs
      name: mc-base-nfs
      pdb:
        currenthealthy: 1
        desiredhealthy: 1
      status: Ready
    - link: /api/v1/namespaces/default/secrets/mc-base-sql
      name: mc-base-sql
      status: Ready
    conditions:
    - lastTransitionTime: 2018-12-07T22:53:23Z
      lastUpdateTime: 2018-12-07T22:53:23Z
      message: all components ready
      reason: ComponentsReady
      status: "True"
      type: Ready
    - lastTransitionTime: 2018-12-07T22:46:00Z
      lastUpdateTime: 2018-12-07T22:46:00Z
      message: No error seen
      reason: NoError
      status: "False"
      type: Error
```

### CODE REVIEW

**Comment by barney-s** [comment]
@kow3ns , @konryd , @mortent , @huyhg - Please review

**Comment by barney-s** [comment]
Willing to iterate the code once the proposals are agreed upon.

File: pkg/apis/app/v1beta1/application.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/kubesdk/pkg/component"
+	cr "sigs.k8s.io/kubesdk/pkg/customresource"
+	"sigs.k8s.io/kubesdk/pkg/finalizer"
+	"sigs.k8s.io/kubesdk/pkg/resource"
+	"sigs.k8s.io/kubesdk/pkg/status"
+	"strings"
+)
+
```
Comment (mortent) [inline pkg/apis/app/v1beta1/application.go]: Some better comments on the functions in this file would be useful. It is unclear to me how each of them fit into the reconcile loop.

File: Makefile [hunk 2]
```diff
@@ -16,6 +16,10 @@ manager: generate fmt vet
 run: generate fmt vet
 	go run ./cmd/manager/main.go
 
+# Run against the configured Kubernetes cluster in ~/.kube/config
+debug: generate fmt vet
+	dlv debug cmd/manager/main.go
```
[the remark this answers is not in the document]
Comment (barney-s) [reply Makefile]: debugger

File: pkg/apis/app/v1beta1/application_types.go [hunk 3]
```diff
@@ -132,8 +131,10 @@ type InfoItem struct {
 	ValueFrom *InfoItemSource `json:"valueFrom,omitempty"`
 }
 
+// InfoItemType is a string
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/apis/app/v1beta1/application_types.go]: done

File: Makefile [hunk 4]
```diff
@@ -16,6 +16,10 @@ manager: generate fmt vet
 run: generate fmt vet
 	go run ./cmd/manager/main.go
 
+# Run against the configured Kubernetes cluster in ~/.kube/config
```
[the remark this answers is not in the document]
Comment (barney-s) [reply Makefile]: changed it to debug

File: pkg/apis/app/v1beta1/application.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	return observed, nil
+}
+
+// Finalize - execute finalizers
+func (a *Application) Finalize(rsrc, sts interface{}, observed *resource.ObjectBag) error {
+	finalizer.Remove(a, finalizer.Cleanup)
+	return nil
+}
+
+//DependantResources - returns dependent rsrc
```
Comment (EronWright) [inline pkg/apis/app/v1beta1/application.go]: Do you mean 'dependent', 'dependant', or both?  The latter is a legit British word.

**Comment by ant31** [comment]
Is it in the state to be merged @barney-s (as the title indicates 'draft') ?

File: config/crds/app_v1beta1_application.yaml [hunk 6]
```diff
@@ -189,6 +189,75 @@ spec:
...
+                properties:
+                  group:
+                    type: string
+                  kind:
+                    type: string
+                  link:
+                    type: string
+                  name:
+                    type: string
+                  pdb:
```
Comment (prydonius) [inline config/crds/app_v1beta1_application.yaml]: Should this be the short name? I think using the full name would make more sense.

Comment (barney-s) [reply config/crds/app_v1beta1_application.yaml]: can change it.

File: pkg/apis/app/v1beta1/application.go (hunk 5 above)
Comment (barney-s) [reply pkg/apis/app/v1beta1/application.go]: will change to "dependent". Had similar comment earlier as well. Guess i am still used to the British spelling.

**Comment by barney-s** [comment]
> Is it in the state to be merged @barney-s (as the title indicates 'draft') ?

Will update the CL to reflect the mergeable status by this week.
Hold merge till then.

File: config/crds/app_v1beta1_application.yaml (hunk 6 above)
Comment (barney-s) [reply config/crds/app_v1beta1_application.yaml]: done

File: pkg/apis/app/v1beta1/application.go (hunk 5 above)
Comment (barney-s) [reply pkg/apis/app/v1beta1/application.go]: done

**Comment by barney-s** [comment]
> > Is it in the state to be merged @barney-s (as the title indicates 'draft') ?
> 
> Will update the CL to reflect the mergeable status by this week.
> Hold merge till then.

Ready for merge.
There are a few Follow up which can be resolved as issues
1. I kept this comment open. Not sure if we need to embed conditions of tracked in app status.
   https://github.com/kubernetes-sigs/application/pull/81#pullrequestreview-193802579
2. shows 2 entries for statefulsets (v1beta2, v1) in status array
3. reconcile loop sometimes resyncs multiple times when single time should be sufficient.

File: pkg/genericreconciler/genericreconciler.go [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+	return e
+}
+
+func (gr *Reconciler) observe(observables ...resource.Observable) (*resource.ObjectBag, error) {
+	var returnval *resource.ObjectBag = new(resource.ObjectBag)
+	var err error
+	for _, obs := range observables {
+		var resources []resource.Object
+		if obs.Labels != nil {
+			//log.Printf("   >>>list: %s labels:[%v]", reflect.TypeOf(obs.ObjList).String(), obs.Labels)
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/genericreconciler/genericreconciler.go]: ack

File: pkg/genericreconciler/genericreconciler.go [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+		log.Printf("%s  Reconciling Resources:\n", cname)
+	}
+	for _, e := range expected.Items() {
+		seen := false
+		eNamespace := e.Obj.GetNamespace()
+		eName := e.Obj.GetName()
+		eKind := reflect.TypeOf(e.Obj).String()
+		eRsrcInfo := eNamespace + "/" + eKind + "/" + eName
+		for _, o := range observed.Items() {
+			if (eName == o.Obj.GetName()) &&
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/genericreconciler/genericreconciler.go]: ack

File: pkg/genericreconciler/v1alpha1/crfoo.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
+/*
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/genericreconciler/v1alpha1/crfoo.go]: removed.

File: pkg/status/types.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	Cleanup = "Cleanup"
+	// Error => last recorded error
+	Error = "Error"
+
+	ReasonInit = "Init"
+)
+
+// Statefulset is a generic status holder for stateful-set
+// +k8s:deepcopy-gen=true
+type Statefulset struct {
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/status/types.go]: These are extended information for core types. Envisioning extending this as we gain more feedback.

File: cmd/manager/main.go [hunk 11]
```diff
@@ -41,8 +42,9 @@ func main() {
 	}
 
 	// Create a new Cmd to provide shared dependencies and start components
+	syncperiod := time.Minute * 2
```
[the remark this answers is not in the document]
Comment (barney-s) [reply cmd/manager/main.go]: just a meaningful default. Alternative is to setup dynamic watchers for all types in all application objects which may quickly escalate to all possible API objects !!

File: pkg/apis/app/v1beta1/application.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+			}
+		}
+
+	}
+	return observables
+}
+
+// Differs returns true if the resource needs to be updated
+func (a *Application) Differs(expected metav1.Object, observed metav1.Object) bool {
+	return true
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/apis/app/v1beta1/application.go]: this is the semantic comparison. true is a lazy implementation punting the update filtering to api server. How the api server does this is something i intend to understand and implement here.

File: pkg/genericreconciler/genericreconciler.go [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+		}
+	}
+	return returnval, nil
+}
+
+func specDiffers(o1, o2 metav1.Object) bool {
+	// Not all k8s objects have Spec
+	// example ConfigMap
+	// TODO strategic merge patch diff in generic controller loop
+	e := reflect.Indirect(reflect.ValueOf(o1)).FieldByName("Spec")
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/genericreconciler/genericreconciler.go]: this is a gate for sematic difference check. if this doesnt pass we call semantic equivalence check via Differs() interface.

File: pkg/genericreconciler/genericreconciler.go [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+	return err
+}
+
+// ObserveAndMutate is a function that is called to observe and mutate expected resources
+func (gr *Reconciler) ObserveAndMutate(crname string, c component.Component, status interface{}, mutate bool, aggregated *resource.ObjectBag) (*resource.ObjectBag, *resource.ObjectBag, string, error) {
+	var err error
+	var expected, observed, dependent *resource.ObjectBag
+
+	// Get dependenta objects
+	stage := "dependent resources"
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/genericreconciler/genericreconciler.go]: done.

File: pkg/genericreconciler/genericreconciler.go [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+	dependent, err = gr.observe(resource.ObservablesFromObjects(gr.Scheme, c.DependentResources(c.CR), c.Labels())...)
+
+	if err == nil && dependent != nil {
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/genericreconciler/genericreconciler.go]: done.

**Comment by barney-s** [comment]
> It would be useful if we had one proposal with consensus on how we want status for Application CRD to work. Based on the three proposals linked here, it is unclear to me how we want this to work.

This is based on #79

File: pkg/status/types.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/status/types.go]: removing for now. to keep it simple.

File: pkg/status/types.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+type Pdb struct {
+	// currentHealthy
+	CurrentHealthy int32 `json:"currenthealthy"`
+	// desiredHealthy
+	DesiredHealthy int32 `json:"desiredhealthy"`
+}
+
+// ExtendedStatus is a holder of additional status for well known types
+// +k8s:deepcopy-gen=true
+type ExtendedStatus struct {
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/status/types.go]: done.

File: pkg/apis/app/v1beta1/application_types.go [hunk 17]
```diff
@@ -78,9 +79,7 @@ type ApplicationSpec struct {
 
 // ApplicationStatus defines controllers the observed state of Application
 type ApplicationStatus struct {
-	// ObservedGeneration is used by the Application Controller to report the last Generation of an Application
-	// that it has observed.
-	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	status.Meta `json:",inline"`
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/apis/app/v1beta1/application_types.go]: status/condition.go methods and UpdateStatus methods are Meta struct methods. 
Extracting UpdateStatus is trivial. But moving all condition functions would need non-trivial refactoring.

Keeping it as a separate package (status) to provide an example for other controllers.

File: pkg/status/types.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
+/*
```
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/status/types.go]: Would renaming Meta to something less confusing help.
Inlining is an artifact of having status code reusable across controllers. (sdk)

**Comment by barney-s** [comment]
> See the with respect to implementing status in its own package and importing it. I think the complexity along with the mirroring of the meta types in api machinery might be confusing. Also its not generally the way that k8s apis are implements.
> You need to fix your code headers.
> So of the code could do with refactoring to a more golang idiomatic happy path style.
> Please remove all commented and uncommented printf debugging statements and all commented out dead code.

- removed dead code
- fixed authors
- refactored for more happy path and less nesting
- removed extended status for known types

with respect to status inline, would renaming meta to something less conflicting help ? Inlining is an artifact of having status package reusable across controllers (sdk)

File: cmd/manager/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (kow3ns) [reply cmd/manager/main.go]: Should probably put that in a comment to make the reasoning more obvious.

File: pkg/status/types.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (kow3ns) [reply pkg/status/types.go]: In this code base there is only one implementation. I don't anticipate anyone vendoring directly from this code base. I don't see why the refactoring would be particularly difficult. If an application sdk is offered as an open source package in the future we should rebase on top of it. In the mean time, I'd really prefer to see all of the relevant types pertaining to application in a single place with a uniform implementation.

[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/status/types.go]: done moving the status bits into application api

File: cmd/manager/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (barney-s) [reply cmd/manager/main.go]: done

File: pkg/apis/app/v1beta1/application_types.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (barney-s) [reply pkg/apis/app/v1beta1/application_types.go]: merged.

**Comment by barney-s** [comment]
@kow3ns 
Moved status bits into application api directly.
Please review and approve.

**Comment by barney-s** [comment]
> Shouldn't the ComponentList contain the group, version, kind and namespace or is the idea that this can be extracted from the link?

A good suggestion. Can do this as a follow up issue.

**Comment by barney-s** [comment]
@huyhg  @mortent @ant31 @prydonius 
Could one of you please review and LGTM ?

File: pkg/apis/app/v1beta1/application_types.go [hunk 19]
```diff
@@ -202,8 +278,10 @@ type IngressSelector struct {
 	Path string `json:"path,omitempty"`
 }
 
+// ApplicationAssemblyPhase tracks the Application CRD phases: pending, succeded, failed
```
Comment (huyhg) [inline pkg/apis/app/v1beta1/application_types.go]: Do we take this into account?

File: pkg/apis/app/v1beta1/status.go [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+		m.NotReady("ComponentsNotReady", "some components not ready")
+	}
+	if err != nil {
+		m.SetCondition(Error, "ErrorSeen", err.Error())
+	}
+}
+
+// Resource specific logic -----------------------------------
+
+// Statefulset
```
Comment (huyhg) [inline pkg/apis/app/v1beta1/status.go]: Just to confirm: more will be added, once the proposal PR is approved, right?

**Comment by kow3ns** [comment]
@huyhg can you lgtm please I already approved

**Comment by huyhg** [comment]
@kow3ns Already LGTM'ed.

**Comment by kow3ns** [comment]
you need to actually "/lgtm" so the label applies and it merges and it has to happen after any commits.

**Comment by cmoulliard** [comment]
Is this PR documented ? Without doc, nobody will be able to use the approach proposed here to handle status ?

**Pull request merged**